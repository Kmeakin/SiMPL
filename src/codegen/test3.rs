use super::{closure::convert, llvm3::Compiler};
use crate::{hir::Expr, types::infer_and_apply};
use inkwell::{context::Context, OptimizationLevel};
use insta::assert_snapshot;
use std::str::FromStr;

#[track_caller]
fn test_compile(src: &str) {
    let expr = Expr::from_str(src).unwrap();
    let expr = infer_and_apply(&expr);
    let cexpr = convert(expr);

    let ctx = Context::create();
    let builder = ctx.create_builder();
    let module = ctx.create_module("test_compile");

    let compiler = Compiler {
        llvm: &ctx,
        module,
        builder,
    };

    let module = compiler.compile_toplevel(&cexpr);

    match module.verify() {
        Ok(()) => {}
        Err(s) => {
            println!("{}\n", module.print_to_string().to_string());
            eprintln!("{}", s.to_string());
            panic!()
        }
    }

    assert_snapshot!(module.print_to_string().to_string());
}

#[track_caller]
fn test_compile_and_execute<T: std::fmt::Debug + PartialEq>(src: &str, expected: T) {
    let expr = Expr::from_str(src).unwrap();
    let expr = infer_and_apply(&expr);
    let cexpr = convert(expr);

    let ctx = Context::create();
    let builder = ctx.create_builder();
    let module = ctx.create_module("test_compile");

    let compiler = Compiler {
        llvm: &ctx,
        module,
        builder,
    };

    let module = compiler.compile_toplevel(&cexpr);

    match module.verify() {
        Ok(()) => {}
        Err(s) => {
            println!("{}\n", module.print_to_string().to_string());
            eprintln!("{}", s.to_string());
            panic!()
        }
    }

    assert_snapshot!(module.print_to_string().to_string());

    let exec_engine = module
        .create_jit_execution_engine(OptimizationLevel::None)
        .unwrap();
    let f = unsafe { exec_engine.get_function::<unsafe extern "C" fn() -> T>("toplevel") }.unwrap();
    assert_eq!(unsafe { f.call() }, expected)
}

#[test]
fn compile_lit() {
    test_compile_and_execute("0", 0);
    test_compile_and_execute("1", 1);
    test_compile_and_execute("true", true);
    test_compile_and_execute("false", false);
    test_compile_and_execute("0.0", 0.0);
    test_compile_and_execute("4.5", 4.5);
}

#[test]
fn compile_vars() {
    test_compile_and_execute("let x = 5 in x", 5);
}

#[test]
fn compile_if() {
    test_compile_and_execute("if true then 5 else 10", 5);
    test_compile_and_execute("if false then 5 else 10", 10);
    test_compile_and_execute(r"let f = \b -> if b then 5 else 10 in f true", 5);
    test_compile_and_execute(r"let f = \b -> if b then 5 else 10 in f false", 10);
}

#[test]
fn compile_lambda() {
    // test_compile(r"\x: Int -> x");
    test_compile(r"let x = 5 in \y: Int -> x");
}

#[test]
fn compile_app() {
    test_compile_and_execute(
        r"
let x = 5,
    f = \y -> x
in f 555
",
        5,
    );
}
