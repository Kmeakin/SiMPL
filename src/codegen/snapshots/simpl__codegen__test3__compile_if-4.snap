---
source: src/codegen/test3.rs
expression: module.print_to_string().to_string()
---
; ModuleID = 'test_compile'
source_filename = "test_compile"

%Closure = type { i8*, i8* }
%Env = type {}

define i64 @toplevel() {
toplevel_entry:
  %f = alloca %Closure
  %closure = alloca %Closure
  %closure.code = getelementptr inbounds %Closure, %Closure* %closure, i32 0, i32 0
  store i8* bitcast (i64 (i8*, i1)* @lambda to i8*), i8** %closure.code
  %closure.env = getelementptr inbounds %Closure, %Closure* %closure, i32 0, i32 1
  %malloccall = tail call i8* @malloc(i32 0)
  %closure.env1 = bitcast i8* %malloccall to %Env*
  %closure.env2 = bitcast %Env* %closure.env1 to i8*
  store i8* %closure.env2, i8** %closure.env
  %closure3 = load %Closure, %Closure* %closure
  store %Closure %closure3, %Closure* %f
  %f4 = load %Closure, %Closure* %f
  %closure5 = alloca %Closure
  store %Closure %f4, %Closure* %closure5
  %closure.fn = getelementptr inbounds %Closure, %Closure* %closure5, i32 0, i32 0
  %closure.fn6 = load i8*, i8** %closure.fn
  %closure.fn7 = bitcast i8* %closure.fn6 to i64 (i8*, i1)*
  %closure.env8 = getelementptr inbounds %Closure, %Closure* %closure5, i32 0, i32 1
  %closure.env9 = load i8*, i8** %closure.env8
  %call = call i64 %closure.fn7(i8* %closure.env9, i1 false)
  ret i64 %call
}

define i64 @lambda(i8* %env, i1 %b) {
lambda_entry:
  %env1 = alloca i8*
  store i8* %env, i8** %env1
  %env2 = load i8*, i8** %env1
  %env3 = bitcast i8* %env2 to %Env*
  %b4 = alloca i1
  store i1 %b, i1* %b4
  %b5 = load i1, i1* %b4
  %cmp = icmp eq i1 %b5, true
  br i1 %cmp, label %then, label %else

then:                                             ; preds = %lambda_entry
  br label %cont

else:                                             ; preds = %lambda_entry
  br label %cont

cont:                                             ; preds = %else, %then
  %phi = phi i64 [ 5, %then ], [ 10, %else ]
  ret i64 %phi
}

declare noalias i8* @malloc(i32)

